### PR TITLE
[#98] We have to allow a user specifying output connection in a train…

### PR DESCRIPTION
**Pull request by vlad-tokarev**
…ing\packaging API.

### CODE REVIEW

File: packages/operator/pkg/apis/packaging/model_packaging.go [hunk 1]
```diff
@@ -42,4 +42,7 @@ type ModelPackagingSpec struct {
 	// Resources for packager container
 	// The same format like k8s uses for pod resources.
 	Resources *v1alpha1.ResourceRequirements `json:"resources,omitempty"`
+	// Name of Connection to storage where packaging output artifact will be stored.
```
Comment (mcdoker18) [inline packages/operator/pkg/apis/packaging/model_packaging.go]: Noop. 
```text
// Name of Connection to storage where a packager obtain a model trained artifact .
```

File: packages/operator/docs/docs.go [hunk 2]
```diff
@@ -70,7 +70,6 @@ var doc = `{
                         "in": "body",
                         "required": true,
                         "schema": {
-                            "type": "object",
```
Comment (mcdoker18) [inline packages/operator/docs/docs.go]: What version of swagger do you use? 
https://github.com/odahu/odahu-flow/blob/develop/.travis.yml#L76

File: helms/odahu-flow-core/templates/operator/generated/odahuflow_v1alpha1_modelpackaging.yaml [hunk 3]
```diff
@@ -39,6 +39,8 @@ spec:
               type: string
             image:
               type: string
+            outputConnection:
+              type: string
```
Comment (mcdoker18) [inline helms/odahu-flow-core/templates/operator/generated/odahuflow_v1alpha1_modelpackaging.yaml]: The documentation is missing.

File: packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_test.go [hunk 4]
```diff
@@ -109,6 +118,30 @@ func (s *ModelPackagingRouteSuite) SetupSuite() {
...
+	s.connRepository = conn_k8s_repository.NewRepository(testNamespace, s.k8sClient, "")
+	// Create the connection that will be used as the outputConnection param for a training.
+	if err := s.connRepository.CreateConnection(&connection.Connection{
+		ID: testOutConn,
+		Spec: odahuflowv1alpha1.ConnectionSpec{
+			Type: connection.GcsType,
+		},
+	}); err != nil {
+		// If we get a panic that we have a test configuration problem
+		panic(err)
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_test.go]: t.Fail

File: packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_test.go [hunk 5]
```diff
@@ -109,6 +118,30 @@ func (s *ModelPackagingRouteSuite) SetupSuite() {
...
+
+	// Create the connection that will be used as the default outputConnection param for a training.
+	if err := s.connRepository.CreateConnection(&connection.Connection{
+		ID: testOutConnDefault,
+		Spec: odahuflowv1alpha1.ConnectionSpec{
+			Type: connection.GcsType,
+		},
+	}); err != nil {
+		// If we get a panic that we have a test configuration problem
+		panic(err)
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_test.go]: t.Fail

File: packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_validation_test.go [hunk 6]
```diff
@@ -465,3 +468,43 @@ func (s *ModelPackagingValidationSuite) TestMpResourcesValidation() {
...
+
+	// If configuration output connection is not set then user must specify it as ModelTraining parameter
+	viper.Set(packaging_config.OutputConnectionName, nil)
+	err := validator.ValidateAndSetDefaults(mp)
+	s.g.Expect(err).To(HaveOccurred())
+	s.g.Expect(err.Error()).To(ContainSubstring(fmt.Sprintf(validation.EmptyValueStringError, "OutputConnection")))
+
+	// If configuration output connection is set and user has not passed output connection as training
+	// parameter then output connection value from configuration will be used as default
+	viper.Set(packaging_config.OutputConnectionName, testMpOutConnDefault)
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/packaging/model_packaging_validation_test.go]: Reset the packaging_config.OutputConnectionName viper value to the nil in the test tear down.

File: packages/operator/pkg/webserver/routes/v1/training/model_training_test.go [hunk 7]
```diff
@@ -94,6 +94,31 @@ func (s *ModelTrainingRouteSuite) SetupSuite() {
...
+		Spec: odahuflowv1alpha1.ConnectionSpec{
+			Type: connection.GcsType,
+		},
+	}); err != nil {
+		// If we get a panic that we have a test configuration problem
+		panic(err)
+	}
+
+	// Define output connection in configuration to not require this from user requests
+	viper.Set(train_config.OutputConnectionName, testMtOutConnDefault)
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/training/model_training_test.go]: Reset the packaging_config.OutputConnectionName viper value to the nil in the test tear down.

File: packages/operator/pkg/webserver/routes/v1/training/model_training_validation.go [hunk 8]
```diff
@@ -23,15 +23,18 @@ import (
...
 	"github.com/odahu/odahu-flow/packages/operator/pkg/repository/util/kubernetes"
+	"github.com/odahu/odahu-flow/packages/operator/pkg/validation"
+	"github.com/spf13/viper"
 	"go.uber.org/multierr"
 	"reflect"
 )
 
 const (
-	mtVcsNotExistsErrorMessage    = "cannot find VCS Credential"
+	MtVcsNotExistsErrorMessage    = "cannot find VCS Credential"
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/training/model_training_validation.go]: connection?

File: packages/operator/pkg/webserver/routes/v1/training/model_training_validation_test.go [hunk 9]
```diff
@@ -409,3 +412,36 @@ func (s *ModelTrainingValidationSuite) TestMtResourcesValidation() {
...
+	mt := &training.ModelTraining{
+		Spec: v1alpha1.ModelTrainingSpec{},
+	}
+	err := s.validator.ValidatesAndSetDefaults(mt)
+	s.g.Expect(err).To(HaveOccurred())
+	s.g.Expect(err.Error()).To(ContainSubstring(fmt.Sprintf(validation.EmptyValueStringError, "OutputConnection")))
+
+	// If configuration output connection is set and user has not passed output connection as training
+	// parameter then output connection value from configuration will be used as default
+	viper.Set(train_config.OutputConnectionName, testMtOutConnDefault)
```
Comment (mcdoker18) [inline packages/operator/pkg/webserver/routes/v1/training/model_training_validation_test.go]: Reset the packaging_config.OutputConnectionName viper value to the nil in the test tear down.

File: packages/operator/docs/docs.go (hunk 2 above)
Comment (vlad-tokarev) [reply packages/operator/docs/docs.go]: (venv) EPUAKHAW1608:operator vlad$ swag --version
swag version v1.6.3

File: helms/odahu-flow-core/templates/operator/generated/odahuflow_v1alpha1_modelpackaging.yaml (hunk 3 above)
Comment (vlad-tokarev) [reply helms/odahu-flow-core/templates/operator/generated/odahuflow_v1alpha1_modelpackaging.yaml]: https://github.com/odahu/odahu-docs/pull/33

**Pull request merged**